src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -794,18 +789,42 @@ private static void finish(@CheckForNull final String cookie) {
         private static final class Callback extends BodyExecutionCallback.TailCall {
 
             private final String cookie;
+            @Deprecated
             private WorkspaceList.Lease lease;
+            private final ExecutorStepExecution execution;
 
-            Callback(String cookie, WorkspaceList.Lease lease) {
+            Callback(String cookie, ExecutorStepExecution execution) {
                 this.cookie = cookie;
-                this.lease = lease;
+                this.execution = execution;
             }
 
             @Override protected void finished(StepContext context) throws Exception {
                 LOGGER.log(FINE, "finished {0}", cookie);
-                lease.release();
-                lease = null;
-                finish(cookie);
+                try {
+                    if (execution != null) {
+                        WorkspaceList.Lease _lease = ExtensionList.lookupSingleton(ExecutorStepDynamicContext.WorkspaceListLeaseTranslator.class).get(execution.state);
```

Comment:
Why can't this be `WorkspaceList.Lease _lease = context.get(WorkspaceList.Lease);`? Because `ExecutorStepDynamicContext` has already been dropped from the context once we enter the callback?

Reply:
Cannot recall. I think this is the wrong context?

Reply:
Checked, and `context.get(WorkspaceList.Lease.class)` is indeed null here. `context` is a `CpsBodySubContext`.